tensorflow/compiler/jit/kernels/xla_ops.cc

```diff
@@ -562,12 +605,26 @@ void XlaRunOp::Compute(OpKernelContext* ctx) {
       ctx, *closure.compilation_result(), closure.num_constant_args());
   OP_REQUIRES_OK(ctx, variable_infos.status());
   OP_REQUIRES_OK(ctx, LockVariables(absl::MakeSpan(*variable_infos)));
+
+  if (cache_ == nullptr) {
```

Comment:
The race seems to be between this line and the assignment of cache_ in rm_->LookupOrCreate.
So I think you need a mutex lock here around this block (and I guess everywhere where cache_ is accessed).